order-book/src/main/java/com/market/banica/order/book/service/grpc/AuroraClient.java

```diff
@@ -0,0 +1,143 @@
+        List<String> startedItems = new ArrayList<>(requestedItems);
+
+        stoppedItems.removeAll(requestedItems);
+        startedItems.removeAll(itemMarket.getItemNameSet());
+
+        startTrackingItems(startedItems, clientId);
+        stopTrackingItems(stoppedItems);
+
+    }
+
+    private void startTrackingItems(List<String> trackedItems, String clientId) throws TrackingException {
+        for (String product : trackedItems) {
+
+            if (cancellableStubs.containsKey(product)) {
+                throw new TrackingException("Item is already being tracked!");
+            }
+
+            final Aurora.AuroraRequest request = Aurora.AuroraRequest.newBuilder()
+                    .setTopic("market/" + product)
+                    .setClientId(clientId)
+                    .build();
+
+            LOGGER.info("Start gathering product data.");
+
+            Context.CancellableContext withCancellation = Context.current().withCancellation();
+            cancellableStubs.put(product, withCancellation);
+            withCancellation.run(() -> startMarketStream(request));
+
+        }
+    }
+
+    private void startMarketStream(Aurora.AuroraRequest request) {
+        final AuroraServiceGrpc.AuroraServiceStub asynchronousStub = AuroraServiceGrpc.newStub(managedChannel);
+
+        asynchronousStub.subscribe(request, new StreamObserver<Aurora.AuroraResponse>() {
+
+            @Override
+            public void onNext(Aurora.AuroraResponse response) {
+                if (response.hasTickResponse()) {
+                    TickResponse tickResponse = response.getTickResponse();
+
+                    Item item = new Item();
+                    item.setPrice(tickResponse.getPrice());
+                    item.setQuantity(tickResponse.getQuantity());
+                    item.setOrigin(tickResponse.getOrigin());
+
+                    itemMarket.getItemSetByName(tickResponse.getGoodName()).add(item);
+
+                    LOGGER.info("Products data updated!");
+                } else {
+                    throw new IncorrectResponseException("Response is not correct!");
+                }
+            }
+
+            @Override
+            public void onError(final Throwable throwable) {
+                LOGGER.warn("Unable to request");
+                LOGGER.error(throwable.getMessage());
+                throwable.printStackTrace();
+            }
+
+            @Override
+            public void onCompleted() {
+                LOGGER.info("Market data gathered");
+            }
+
+        });
+
+    }
+
+    private void stopTrackingItems(List<String> untrackedItems) throws TrackingException {
+        for (String product : untrackedItems) {
+
+            if (!cancellableStubs.containsKey(product)) {
+                throw new TrackingException("Item is not being tracked!");
+            }
+
+            Context.CancellableContext cancelledStub = cancellableStubs.remove(product);
+            cancelledStub.cancel(null);
```

Comment:
Please provide information about why it's cancelled

Reply:
Added an exception and threw when cancelling a stream.